examples/cpp/waku.cpp

```diff
@@ -0,0 +1,336 @@
+	size_t  v;
+
+	if (in == NULL || len == 0)
+		return;
+
+	elen = b64_encoded_size(len);
+	out.reserve(elen+1);
+
+	for (i=0, j=0; i<len; i+=3, j+=4) {
+		v = in[i];
+		v = i+1 < len ? v << 8 | in[i+1] : v << 8;
+		v = i+2 < len ? v << 8 | in[i+2] : v << 8;
+
+		out[j]   = b64chars[(v >> 18) & 0x3F];
+		out[j+1] = b64chars[(v >> 12) & 0x3F];
+		if (i+1 < len) {
+			out[j+2] = b64chars[(v >> 6) & 0x3F];
+		} else {
+			out[j+2] = '=';
+		}
+		if (i+2 < len) {
+			out[j+3] = b64chars[v & 0x3F];
+		} else {
+			out[j+3] = '=';
+		}
+	}
+}
+
+// End of Base64 encoding
+
+// Beginning of UI program logic
+
+enum PROGRAM_STATE {
+    MAIN_MENU,
+    SUBSCRIBE_TOPIC_MENU,
+    CONNECT_TO_OTHER_NODE_MENU,
+    PUBLISH_MESSAGE_MENU
+};
+
+enum PROGRAM_STATE current_state = MAIN_MENU;
+
+void show_main_menu() {
+    printf("\nPlease, select an option:\n");
+    printf("\t1.) Subscribe to topic\n");
+    printf("\t2.) Connect to other node\n");
+    printf("\t3.) Publish a message\n");
+}
+
+void set_scanf_to_not_block() {
+    fcntl(0, F_SETFL, fcntl(0, F_GETFL) | O_NONBLOCK);
+}
+
+void set_scanf_to_block() {
+    fcntl(0, F_SETFL, fcntl(0, F_GETFL) ^ O_NONBLOCK);
+}
+
+void handle_user_input() {
+    char cmd[1024];
+    memset(cmd, 0, 1024);
+    int numRead = read(0, cmd, 1024);
+    if (numRead <= 0) {
+        return;
+    }
+
+    int c;
+    while ( (c = getchar()) != '\n' && c != EOF ) { }
+
+    switch (atoi(cmd))
+    {
+    case SUBSCRIBE_TOPIC_MENU:
+    {
+        printf("Indicate the Pubsubtopic to subscribe:\n");
+        set_scanf_to_block();
+        char pubsubTopic[128];
+        scanf("%127s", pubsubTopic);
+        // if (!waku_relay_subscribe(pubsubTopic, &mResp)) {
+        //     printf("Error subscribing to PubsubTopic: %s\n", mResp->data);
+        // }
+        // printf("Waku Relay subscription response: %s\n", mResp->data);
+
+        set_scanf_to_not_block();
+        show_main_menu();
+    }
+    break;
+
+    case CONNECT_TO_OTHER_NODE_MENU:
+        printf("Connecting to a node. Please indicate the peer Multiaddress:\n");
+        printf("e.g.: /ip4/127.0.0.1/tcp/60001/p2p/16Uiu2HAmVFXtAfSj4EiR7mL2KvL4EE2wztuQgUSBoj2Jx2KeXFLN\n");
+        set_scanf_to_block();
+        char peerAddr[512];
+        scanf("%511s", peerAddr);
+        // if (!waku_connect(peerAddr, 10000 /* timeoutMs */, &mResp)) {
+        //     printf("Couldn't connect to the remote peer: %s\n", mResp->data);
+        // }
+        set_scanf_to_not_block();
+        show_main_menu();
+    break;
+
+    case PUBLISH_MESSAGE_MENU:
+    {
+        set_scanf_to_block();
+        printf("Indicate the Pubsubtopic:\n");
+        char pubsubTopic[128];
+        scanf("%127s", pubsubTopic);
+
+        printf("Type the message tp publish:\n");
+        char msg[1024];
+        scanf("%1023s", msg);
+
+        char jsonWakuMsg[1024];
+        std::vector<char> msgPayload;
+        b64_encode(msg, strlen(msg), msgPayload);
+
+        // waku_content_topic("appName",
```

Comment:
Is it something intentionally left commented our not working?

Reply:
Good catch! I leave those comments to indicate which `libwaku` should be invoked. If we uncomment they might not work because the signatures are changing due to making the `libwaku` thread-safe. We should revisit that once we have time to tackle it.